### PR TITLE
Correct Annotations

### PR DESCRIPTION
**Pull request by sHermanGriffiths**
Spaces at the end and beginning of certain annotations would cause a glitch in the conversion to markdown. The code added in this branch is meant to correct those glitches

### CODE REVIEW

File: tests/test_converter.py
Comment (johndgiese) [inline tests/test_converter.py:579]: Leftover print statement?

File: tests/test_converter.py
Comment (johndgiese) [inline tests/test_converter.py:418]: Leftover print statement?

File: n2y/converter.py
Comment (johndgiese) [inline n2y/converter.py:175]: I'll leave it to @xaker00 to review this algorithm in detail, but it looks nice at first glance!

File: tests/test_converter.py
Comment (johndgiese) [inline tests/test_converter.py:463]: Is it necessary to include all of this meta data, e.g., `created_time`, in the test fixture here? It seems unrelated to what we're testing. If it is needed, perhaps there's some way we can create a mock utility so that we can hide this complexity outside of the tests. If other tests include this metadata unnecessarily, I suggest we refactor those other tests as well.

Comment (sHermanGriffiths) [reply tests/test_converter.py:462]: no, I can certainly and will delete some of it

File: tests/test_converter.py
Comment (sHermanGriffiths) [reply tests/test_converter.py]: yea, there were a bunch haha

File: tests/test_converter.py
Comment (johndgiese) [inline tests/test_converter.py:468]: This is a big object to read through to understand what the test is doing. What do you think about this idea:

Write a utility function that can generate the objects for us using an easier to read format. I've used this technique often in other tests:

```python
input = annotated_text([("this ", "b"), ("is ", "bi"), ("a", "I"), (" test", "s")])
```

I think you get the idea.

Also, although I didn't read through the test data carefully, I suspect you're testing a few different things all in one big test. I suspect you did this in part because the current test data fixture is so big it felt bad to copy it a few times for multiple tests. Once you've created this little function I showed you, you can split apart this big test into a few smaller tests that all have good function names. Each test should test one particular situation. You may need several of them to cover all the corner cases.

Let me know if any of this is unclear and I'd be happy to help.

Comment (johndgiese) [reply tests/test_converter.py:468]: I'd also update the other tests to use this new tool, where appropriate.

Comment (sHermanGriffiths) [reply tests/test_converter.py:468]: makes sense, I'll comment when it's done

Comment (johndgiese) [reply tests/test_converter.py]: Very close, nice work! everything you've added looks great but please reread my comment carefully and be sure you addressed everything I brought up. Also, I think you can use tuple unpacking to in the for loop so that you can give block[0] and block[1] better names.

File: tests/test_converter.py
Comment (johndgiese) [inline tests/test_converter.py:31]: You can expand the tuple in the for loop directly!

Also, `arr` is not a very good name. I'd change it to something else, like `text_blocks_descriptors`:

```suggestion
    for (text, annotations) in text_blocks_descriptors:
```

**Review by johndgiese — APPROVED** [review]
@sHermanGriffiths I had one small suggested changed which I'd ask you to make, but otherwise this looks great! I'll defer to @xaker00 if he wants to review the algorithm itself.

File: n2y/converter.py
Comment (xaker00) [inline n2y/converter.py:182]: Logic error here. `prependages` ends up reversed here

File: n2y/converter.py
Comment (xaker00) [inline n2y/converter.py:175]: `strip` does more than strip. It does the conversion also.

**Review by xaker00 — CHANGES_REQUESTED** [review]
Good job improving and cleaning up tests. I would suggest a simpler version of the `Annotations` class:

```python
class Annotations():
    def __init__(self, block):
        for key, value in block.items():
            self.__dict__[key] = value

    def apply_pandoc(self, target):
        prependages = deque()
        appendages = deque()

        blank_space = [Space(), SoftBreak()]
        while target[0] in blank_space:
            prependages.append(target.pop(0))
        while target[-1] in blank_space:
            appendages.appendleft(target.pop(-1))

        def listify(obj):
            return obj if type(obj) == list else [obj]

        result = target
        if self.code:
            result = listify(Code(("", [], []), result))
        if self.bold:
            result = listify(Strong(result))
        if self.italic:
            result = listify(Emph(result))
        if self.underline:
            result = listify(Underline(result))
        if self.strikethrough:
            result = listify(Strikeout(result))

        return list(prependages) + result + list(appendages)
```

File: n2y/converter.py
Comment (sHermanGriffiths) [reply n2y/converter.py:182]: it's actually on purpose. I wanted to use the deque like a stack. I figure this way it left appends and now the first thing served in the for loop is the last thing added.

Comment (sHermanGriffiths) [reply n2y/converter.py:182]: Thanks, I was visualizing it incorrectly. The appendages _should_ be appended left though, right?

File: n2y/converter.py
Comment (sHermanGriffiths) [reply n2y/converter.py:201]: Yes but only the bold, italic, and strikethrough annotations can't affect leading and trailing spaces in gfm so I figure only change those in order to make our conversion as close to the original document as possible, only making changes when absolutely neccessary.

Comment (johndgiese) [reply n2y/converter.py]: I like the idea of making the minimal number of changes to the original document as possible. We may not even need to trim the space for strikethrough _if_ struck out spaces are visible in Notion. I think we definitely need to handle bold and italic because bolded spaces look just like unbolded spaces and italicized spaces look just like unitalicized spaces.

File: n2y/converter.py
Comment (johndgiese) [inline n2y/converter.py:1]: What is this import for?

File: n2y/converter.py
Comment (xaker00) [reply n2y/converter.py]: You have two loops one going forward, and the other backward. When you reconstruct the result in the end, all of the pieces need to be arranged left to right. This is why on the reverse loop the items need to be added to the other side. In effect reversing a reversed list and making it correct.

File: n2y/converter.py
Comment (xaker00) [reply n2y/converter.py]: leading and trailing spaces confuse markdown renderers. In the case of italic, a leading space will add a bullet instead of making the word italic.

This is how vscode renders:
image

File: n2y/converter.py
Comment (xaker00) [reply n2y/converter.py:1]: That is vscode trying to be helpful. Have to watch out for these. Please delete @sHermanGriffiths.

In vscode if you try to use a variable without first assigning it, vscode will try to find something that it can import to match that name. This is both helpful and annoying. Vscode is awesome for some things, but not as smart as PyCharm when it comes to python.

**Review by xaker00 — APPROVED** [review]
Looks good. Can be merged after removing the extra import mentioned here: https://github.com/innolitics/n2y/pull/18#discussion_r841739491.

File: n2y/converter.py
Comment (sHermanGriffiths) [reply n2y/converter.py:1]: deleted

File: n2y/converter.py
Comment (sHermanGriffiths) [reply n2y/converter.py]: > We may not even need to trim the space for strikethrough _if_ struck out spaces are visible in Notion.

@johndgiese I know that Notion can render struck out spaces just fine but the final markdown conversion renders as:
 ~~ STRIKETHROUGH ~~
Instead of something that looks like:
-~~STRIKETHROUGH~~-

With that in mind, I think we should still correct them. What do you think?

**Comment by johndgiese** [comment]
Agreed!

**Pull request merged**
